### PR TITLE
fix #6, escape (possibly) `throw`n arguments

### LINKED ISSUE #6: recognize escapes from implicit `throw`s involved with builtins
MRE:
```julia
julia> analyze_escapes((Type{Array},)) do t
           sizeof(t)
       end
2 1 ─ %1 = Core.sizeof::typeof(Core.sizeof)                                                                                                                                                                          │╻ sizeof
  │   %2 = (%1)(_2)::Int64                                                                                                                                                                                           ││
  └──      return %2                                                                                                                                                                                                 │ 
  , EscapeAnalysis.EscapeState(                                                                                                                                                                                        
EscapeAnalysis.EscapeInformation[EscapeAnalysis.ReturnEscape(), EscapeAnalysis.ReturnEscape()], EscapeAnalysis.EscapeInformation[EscapeAnalysis.NoInformation(), EscapeAnalysis.ReturnEscape(), EscapeAnalysis.NoInformation()]))
```

Actually `sizeof(t)` throws in this case, and thus `t` escapes as an exception.
We may want to use [the existing query to check `throw`-freeness](https://github.com/JuliaLang/julia/blob/71bf60858ee2dcf219f6de568f712a0ab3756549/base/compiler/ssair/queries.jl#L6).

### PR DESCRIPTION
**Pull request by aviatesk**
Currently this PR just re-uses ["effect-freeness" of a statement](https://github.com/JuliaLang/julia/blob/94b9d66b10e8e3ebdb268e4be5f7e1f43079ad4e/base/compiler/ssair/queries.jl#L6)
computed by inliner.

This is necessary but we lose much analysis accuracy, e.g. now we can't
recognize the target below (i.e. `%1` escapes):
```julia
julia> @noinline f_noescape(x) = (broadcast(identity, x); nothing);

julia> analyze_escapes() do
           f_noescape(Ref("Hi"))
       end
var"#3#4"()
1 → 1 ─ %1 = %new(Base.RefValue{String}, "Hi")::Base.RefValue{String}
2 ↑ │   %2 = invoke Main.f_noescape(%1::Base.RefValue{String})::Core.Const(nothing)
3 ◌ └──      return %2
```

This case is particularly because Julia compiler doesn't tell
`getfield(::Base.RefValue, :x)` is "effect-free" since it's caught by
[this `field <= datatype_min_ninitialized(s)` check](https://github.com/JuliaLang/julia/blob/94b9d66b10e8e3ebdb268e4be5f7e1f43079ad4e/base/compiler/tfuncs.jl#L745).
We may need to improve inference instead rather than doing something in
this pass.

### CODE REVIEW

File: src/EscapeAnalysis.jl [hunk 1]
```diff
@@ -207,39 +208,49 @@ function find_escapes(ir::IRCode, nargs::Int)
...
-                    elseif ft === typeof(ifelse) && length(stmt.args) === 4
-                        f, cond, th, el = stmt.args
-                        info = state.ssavalues[pc]
-                        condt = argextype(cond, ir, sptypes, argtypes)
-                        if isa(condt, Const) && isa(condt.val, Bool)
-                            if condt.val
-                                push!(changes, (th, info))
+                if head === :call # TODO implement more builtins, make them more accurate
+                    if !is_effect_free
+                        # TODO we can have a look at builtins.c and limit the escaped arguments if any of them are not thrown
```
Comment (TH3CHARLie) [inline src/EscapeAnalysis.jl:220]: I am not very sure about this TODO, do you mean that we can prevent propagating unnecessary escapes if we find the builtin function is not throwing, which means the `IR_FLAG_EFFECT_FREE ` can be set by events other than `throw`?

**Review by TH3CHARLie — APPROVED** [review]
Thanks for the update and looks good to me. I left one comment about my confusion.

File: src/EscapeAnalysis.jl (hunk 1 above)
Comment (aviatesk) [reply src/EscapeAnalysis.jl:220]: I meant we don't necessarily escape all the arguments here, e.g. `isa(o, t)` doesn't escape `o` even when it `throw`s.

Comment (TH3CHARLie) [reply src/EscapeAnalysis.jl:220]: I see, thanks for the explanation

**Comment by TH3CHARLie** [comment]
I believe we've talked about this during the meeting, for builtin like `Core.arraysize`, it is marking args as escaped because it is not effect-free. But to me, this method shouldn't escape the arg. Should we special case this as well?

**Comment by aviatesk** [comment]
> But to me, this method shouldn't escape the arg.

Why do you think so ?

**Comment by TH3CHARLie** [comment]
> > But to me, this method shouldn't escape the arg.
> 
> Why do you think so ?

https://github.com/JuliaLang/julia/blob/42d810559d5bc0d9a4cd5b99b6b449e24faf2780/src/builtins.c#L1330-L1343 In this implementation, apart from the throw, it is just reading a bitstype field. I think this may related to your TODO comment here about "we can have a look at builtins.c and limit the escaped arguments if any of them are not thrown "

**Comment by aviatesk** [comment]
When `throw`n, arguments escape through `JL_TYPECHK`, no ?

**Comment by TH3CHARLie** [comment]
OK, that's a fair point..... I forget `JL_TYPECHK` escapes. Not sure if the potential new lattice element `MayEscape` fits here.

**Comment by aviatesk** [comment]
The highest priority around this problem would be put on #15.

**Pull request merged**